File: ruby/palindrome-products/palindrome_products.rb
require 'set'

# Palindrome products
class Palindromes
  def initialize(options = {})
    @min = options[:min_factor] || 0
    @max = options[:max_factor] || 0
    @palindromes = {}
  end

  def generate
    (@min..@max).to_a.product((@min..@max).to_a).each do |x, y|
      value = x * y

      next unless palindrome?(value) || value == 0

      palindrome = @palindromes.fetch(value, Palindrome.new(value))

      palindrome.factor_set.add(x <= y ? [x, y] : [y, x])

      @palindromes[value] = palindrome
    end
  end

  def palindrome?(n)
    n.to_s.reverse == n.to_s
  end

  def largest
    @palindromes[@palindromes.keys.max]
  end

  def smallest
    @palindromes[@palindromes.keys.min]
  end
end

# Palindrome with factors
class Palindrome
  include Comparable

  attr_accessor :value, :factor_set

  def initialize(value)
    @value = value
    @factor_set = Set.new
  end

  def factors
    @factor_set.to_a
  end
end

File: elixir/palindrome-products/palindrome_products.exs
defmodule Palindromes do

  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map
  def generate(max_factor, min_factor \\ 1) do
    (for x <- min_factor..max_factor, y <- min_factor..max_factor, do: Enum.sort([x, y]))
    |> Enum.filter(fn [x, y] -> palindrome?(x * y) end)
    |> Enum.reduce(%{}, &update_map/2)
    |> Enum.into(%{}, &dedupe_list/1)
  end

  defp dedupe_list({k, v}), do: {k, v |> MapSet.new |> MapSet.to_list}
  defp update_map([x, y], map), do: Map.update(map, x * y, [[x, y]], &[[x, y] | &1])

  defp palindrome?(str) when is_binary(str), do: str == String.reverse(str)
  defp palindrome?(num) when is_number(num), do: palindrome?(to_string(num))
end
